Fix app can quit while playback history is being saved, #5107

low-batt opened the pull request:
This commit will:
- Change the call to `write` in the `HistoryController.save` method to use the `atomic` option
- Add a` tasksOutstanding` property to `HistoryController` to keep track of how many tasks are in the background queue
- Change the `HistoryController.history` property to be atomic
- Change the methods in `HistoryController`, `HistoryWindowController` and `JavascriptAPICore` to lock the `history` property when accessing it
- Add a `iinaHistoryTaskFinished` notification to `AppData`
- Change `HistoryController.add` to post this notification when a background task finishes
- Change `AppDelegate.applicationShouldTerminate` to establish an observer for the `iinaHistoryTaskFinished` notification and wait until all outstanding `HistoryController` tasks have finished before proceeding with termination
- Add logging to the `HistoryController`
- Remove unused `removeAll` method from `HistoryController`

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #5107.

---

**Description:**

low-batt committed:
Fix app can quit while playback history is being saved, #5107

This commit will:
- Change the call to write in the HistoryController.save method to use
  the atomic option
- Add a tasksOutstanding property to HistoryController to keep track of
  how many tasks are in the background queue
- Change the HistoryController.history property to be atomic
- Change the methods in HistoryController, HistoryWindowController and
  JavascriptAPICore to lock the history property when accessing it
- Add a iinaHistoryTaskFinished notification to AppData
- Change HistoryController.add to post this notification when a
  background task finishes
- Change AppDelegate.applicationShouldTerminate to establish an observer
  for the iinaHistoryTaskFinished notification and wait until all
  outstanding HistoryController tasks have finished before proceeding
  with termination
- Add logging to the HistoryController
- Remove unused removeAll method from HistoryController

lhc70000 commented on iina/AppDelegate.swift:
```diff
@@ -573,13 +577,16 @@ class AppDelegate: NSObject, NSApplicationDelegate, SPUUpdaterDelegate {
-      // All players have shutdown. No longer logged into an online subtitles provider.
+      // If still logged in to subtitle providers or still saving playback history then continue
+      // waiting.
+      guard !OnlineSubtitle.loggedIn, HistoryController.shared.tasksOutstanding == 0 else { return }
```
I suggest using two separate guards.

low-batt replied:
Fixed. I initially had it as two guards as I thought that was easier to read with separate comments. I decided to combine it into one guard right before posting as a lot of the code uses guards with multiple expressions. In this case with the separate guards it is quicker to read and understand.

lhc70000 reviewed (commented):
Looks good to me.

I was thinking whether an `OperationQueue` will be more suitable for this case, but all other checks in `proceedWithTermination()` use the notification based mechanism. So the current implementation is fine.

low-batt committed:
Fix app can quit while playback history is being saved, #5107

This commit will:
- Change the call to write in the HistoryController.save method to use
  the atomic option
- Add a tasksOutstanding property to HistoryController to keep track of
  how many tasks are in the background queue
- Change the HistoryController.history property to be atomic
- Change the methods in HistoryController, HistoryWindowController and
  JavascriptAPICore to lock the history property when accessing it
- Add a iinaHistoryTaskFinished notification to AppData
- Change HistoryController.add to post this notification when a
  background task finishes
- Change AppDelegate.applicationShouldTerminate to establish an observer
  for the iinaHistoryTaskFinished notification and wait until all
  outstanding HistoryController tasks have finished before proceeding
  with termination
- Add logging to the HistoryController
- Remove unused removeAll method from HistoryController

low-batt commented:
I too considered `OperationQueue` as well as [DispatchWorkItem](https://developer.apple.com/documentation/dispatch/dispatchworkitem) but I didn't see an advantage to using them over the current event driven system.
